File: leetcode/problems/0003-longest-substring-without-repeating-characters/0003.js
/**
 * Runtime: 140 ms
 * Memory Usage: 44.5 MB
 */
 var lengthOfLongestSubstring = function(s) {
  let ans = 0, start = 0;
  
  const map = {};

  for(let i = 0; i < s.length; i++)
  {
      let c = s[i];
      if(map[c] !== undefined){
          start = Math.max(start, map[c] + 1);
      }
      ans = Math.max(ans, i-start+1);
      
      map[c] = i;
  }

  return ans;
};

File: leetcode/problems/0003-longest-substring-without-repeating-characters/0003.java
import java.util.HashMap;

/**
 * Runtime: 4 ms
 * Memory Usage: 38.6 MB
 */
class Solution {
  public int lengthOfLongestSubstring(String s) {
      int ans = 0;
      int start = 0;
      
      HashMap<Character, Integer> map = new HashMap<>();
      
      for (int i = 0; i < s.length(); i++) {
          char c = s.charAt(i);
          
          if (map.containsKey(c)) {
              start = Math.max(start, map.get(c) + 1);
          }
          ans = Math.max(ans, i - start + 1);
          map.put(c, i);
      }
      return ans;
  }
}